hydroflow/tests/surface_codegen.rs

```diff
@@ -71,27 +71,23 @@ pub fn test_surface_syntax_reachability_generated() {
 
 #[test]
 pub fn test_transitive_closure() {
-    // WIP
-
     // An edge in the input data = a pair of `usize` vertex IDs.
     let (pairs_send, pairs_recv) = tokio::sync::mpsc::unbounded_channel::<(usize, usize)>();
 
     let mut df = hydroflow_syntax! {
-        reached_vertices = (merge() -> map(|v| (v, ())));
-
-        input_tee = tee();
-        node_merge = merge();
-        (input(pairs_recv) -> input_tee);
-        (input_tee[0] -> map(|v: (usize, usize)| v.0) -> [0]node_merge);
-        (input_tee[1] -> map(|v: (usize, usize)| v.1) -> [1]node_merge);
-        (node_merge -> [0]reached_vertices);
-
-        my_join_tee = (join() -> map(|(_src, ((), dst))| dst) -> tee());
-        (reached_vertices -> [0]my_join_tee);
-        (input_tee[2] -> [1]my_join_tee);
-
-        (my_join_tee[0] -> [1]reached_vertices);
-        (my_join_tee[1] -> for_each(|x| println!("Reached: {}", x)));
+        // edge(x,y) :- link(x,y)
+        edge_merge_tee = (merge() -> tee());
+        link_tee = tee();
+        (input(pairs_recv) -> link_tee);
+        (link_tee[0] -> [0]edge_merge_tee);
+
+        // edge(x,y) :- edge(x,y), link(y,z)
+        the_join = join();
+        (edge_merge_tee[0] -> map(|x: (usize, usize)| (x.1, (x.0,x.1))) -> [0]the_join);
+        (link_tee[1] -> map(|x: (usize, usize)| (x.0, (x.0, x.1))) -> [1]the_join);
+        // the_join emits (x, ((x,y), (y,z)))
+        (the_join -> map(|t: (usize, ((usize, usize), (usize, usize)))| (t.1.0.0, t.1.1.1)) -> [1]edge_merge_tee);
+        (edge_merge_tee[1] -> for_each(|x: (usize, usize)| println!("transitive closure: ({},{})", x.0, x.1)));
```

Comment:
Can simplify this:
```rs
        (edge_merge_tee[0] -> map(|(a, k)| (k, a)) -> [0]the_join);
        (link_tee[1] -> map(|(k, b)| (k, b)) -> [1]the_join);
        (the_join -> map(|(_k, (a, b))| (a, b)) -> [1]edge_merge_tee);
        (edge_merge_tee[1] -> for_each(|(a, b)| println!("transitive closure: ({},{})", a, b)));
```

Syntax like `|(a, b)| (b, a)` is shorthand for `|t| (t.1, t.0)`. This is called "destructuring"

Reply:
nice. can clean up further -- link_tee[1] doesn't need mapping.